### PR TITLE
update to NIO 2

### PR DESCRIPTION
**Pull request by tanner0101**
I needed this fork to use OpenSSL with NIO 2 (master branch). Contributing back here just in case you guys might want to use it.

### CODE REVIEW

**Comment by Lukasa** [comment]
We don't have CI for this yet, and we haven't cut a maintenance branch yet. I'll work on starting that but realistically I don't expect it to complete until the new year.

**Comment by pushkarnk** [comment]
@Lukasa @weissi When could we have this merged? We plan to maintain a development branch for Kitura-NIO using NIO's master

**Comment by pushkarnk** [comment]
I guess this PR must include changes to Tests too?

**Comment by Lukasa** [comment]
@pushkarnk We can't merge it with outstanding notes in the code review. :smile:

With that said, I want to push a new release of this repository sometime shortly, and when I do so I'll probably take advantage of that time to branch off a maintenance branch and convert `master` to Swift 5 and NIO 2.

**Comment by pushkarnk** [comment]
@Lukasa sorry, I missed the `Changes Requested` tag initially. Sure, your plan sounds good. Thanks!

Most of the [needed changes](https://github.com/pushkarnk/swift-nio-ssl/tree/master) are trivial. I guess the only change that will need some work is around `ClientSNITests`, owing to the `SNIHandler` not  conforming to `ChannelHandler` anymore.

**Comment by tanner0101** [comment]
Tools version updated to `5.0` 👍

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by Lukasa** [comment]
@tomerd I don't think this actually worked. Specifically, it seems that the 5.0 test is still installing Swift 4.0.3.

**Comment by tomerd** [comment]
@Lukasa why do you think that? it issuing the following command which is ubuntu 18.04 and swift 5.0 afaict, you should be able to run this locally too

`docker-compose -f docker/docker-compose.yaml -f docker/docker-compose.1804.50.yaml -p swift-nio-ssl2-swift50-prb run --rm test`

**Comment by tomerd** [comment]
build error is: 

`00:00:06.364 /code: error: package at '/code' is using Swift tools version 4.2.0 which is no longer supported; use 5.0.0 or newer instead`

**Comment by Lukasa** [comment]
Yup, but I think that error message is confusingly backwards. I think it means that the *actual* swift tools version is 4.2.0. That's because this patch adds a requirement to use swift tools version 5.

**Comment by tomerd** [comment]
@Lukasa from everything the docker compose setup is fine and using swift 5. here is a repro for your local machine:

```
tomerd@tombp swift-nio-ssl (master) $ docker-compose -f docker/docker-compose.yaml -f docker/docker-compose.1804.50.yaml build 
common uses an image, skipping
test uses an image, skipping
sanity uses an image, skipping
Building runtime-setup
Step 1/38 : ARG ubuntu_version=16.04
Step 2/38 : FROM ubuntu:$ubuntu_version
 ---> cd6d8154f1e1
Step 3/38 : ARG ubuntu_version
 ---> Using cache
 ---> 968674528d60
Step 4/38 : ARG DEBIAN_FRONTEND=noninteractive
 ---> Using cache
 ---> 1bc155744aa8
Step 5/38 : RUN echo "#!/bin/sh\nexit 0" > /usr/sbin/policy-rc.d
 ---> Using cache
 ---> 6e53fded0065
Step 6/38 : RUN apt-get update && apt-get install -y wget git build-essential software-properties-common pkg-config locales
 ---> Using cache
 ---> d00bde0007ad
Step 7/38 : RUN apt-get update && apt-get install -y libicu-dev libblocksruntime0 libssl-dev
 ---> Using cache
 ---> 7dd444e79327
Step 8/38 : RUN locale-gen en_US.UTF-8
 ---> Using cache
 ---> de6d9e398451
Step 9/38 : RUN locale-gen en_US en_US.UTF-8
 ---> Using cache
 ---> 4f82c7b44fe0
Step 10/38 : RUN dpkg-reconfigure locales
 ---> Using cache
 ---> 7eee8f57e107
Step 11/38 : RUN echo 'export LANG=en_US.UTF-8' >> $HOME/.profile
 ---> Using cache
 ---> c268ad0c0c85
Step 12/38 : RUN echo 'export LANGUAGE=en_US:en' >> $HOME/.profile
 ---> Using cache
 ---> d9e91f42111b
Step 13/38 : RUN echo 'export LC_ALL=en_US.UTF-8' >> $HOME/.profile
 ---> Using cache
 ---> 4850c062cbc9
Step 14/38 : RUN mkdir -p $HOME/.ssh
 ---> Using cache
 ---> 526602def701
Step 15/38 : RUN touch $HOME/.ssh/known_hosts
 ---> Using cache
 ---> c59868095324
Step 16/38 : RUN ssh-keyscan github.com 2> /dev/null >> $HOME/.ssh/known_hosts
 ---> Using cache
 ---> 9c72748e2fbc
Step 17/38 : RUN apt-get update && apt-get install -y clang-3.9
 ---> Using cache
 ---> 936bbed8e324
Step 18/38 : RUN update-alternatives --install /usr/bin/clang clang /usr/bin/clang-3.9 100
 ---> Using cache
 ---> 926044134bb9
Step 19/38 : RUN update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-3.9 100
 ---> Using cache
 ---> 50d07d3d2d40
Step 20/38 : ARG install_curl_from_source
 ---> Using cache
 ---> 5e0f3b8b5fde
Step 21/38 : RUN apt-get update && apt-get install -y curl libcurl4-openssl-dev libz-dev
 ---> Using cache
 ---> 703c6f0c7dbd
Step 22/38 : ARG skip_ruby_from_ppa
 ---> Using cache
 ---> 35383f05fcfe
Step 23/38 : RUN [ -n "$skip_ruby_from_ppa" ] || apt-add-repository -y ppa:brightbox/ruby-ng
 ---> Using cache
 ---> 17b71beeda1a
Step 24/38 : RUN [ -n "$skip_ruby_from_ppa" ] || { apt-get update && apt-get install -y ruby2.4 ruby2.4-dev; }
 ---> Using cache
 ---> 748cfc490a0d
Step 25/38 : RUN [ -z "$skip_ruby_from_ppa" ] || { apt-get update && apt-get install -y ruby ruby-dev; }
 ---> Using cache
 ---> 294471fef752
Step 26/38 : RUN apt-get update && apt-get install -y libsqlite3-dev
 ---> Using cache
 ---> dabaa8c23a7f
Step 27/38 : ARG swift_version=4.0.3
 ---> Using cache
 ---> 661cd9c475b8
Step 28/38 : ARG swift_flavour=RELEASE
 ---> Using cache
 ---> 2b0488819800
Step 29/38 : ARG swift_builds_suffix=release
 ---> Using cache
 ---> 0431703796af
Step 30/38 : RUN mkdir $HOME/.swift
 ---> Using cache
 ---> 4da010153bf4
Step 31/38 : RUN wget -q "https://swift.org/builds/swift-${swift_version}-${swift_builds_suffix}/ubuntu$(echo $ubuntu_version | sed 's/\.//g')/swift-${swift_version}-${swift_flavour}/swift-${swift_version}-${swift_flavour}-ubuntu${ubuntu_version}.tar.gz" -O $HOME/swift.tar.gz
 ---> Using cache
 ---> 788e3cc6edbd
Step 32/38 : RUN tar xzf $HOME/swift.tar.gz --directory $HOME/.swift --strip-components=1
 ---> Using cache
 ---> a117ba921ca0
Step 33/38 : RUN echo 'export PATH="$HOME/.swift/usr/bin:$PATH"' >> $HOME/.profile
 ---> Using cache
 ---> 7946387f5078
Step 34/38 : RUN echo 'export LINUX_SOURCEKIT_LIB_PATH="$HOME/.swift/usr/lib"' >> $HOME/.profile
 ---> Using cache
 ---> a501a78ded82
Step 35/38 : RUN mkdir -p $HOME/.scripts
 ---> Using cache
 ---> dcefe66d64d1
Step 36/38 : RUN wget -q https://raw.githubusercontent.com/apple/swift/master/utils/symbolicate-linux-fatal -O $HOME/.scripts/symbolicate-linux-fatal
 ---> Using cache
 ---> d4e65e701698
Step 37/38 : RUN chmod 755 $HOME/.scripts/symbolicate-linux-fatal
 ---> Using cache
 ---> bdf3d6f5e7ca
Step 38/38 : RUN echo 'export PATH="$HOME/.scripts:$PATH"' >> $HOME/.profile
 ---> Using cache
 ---> 1b1278345097
Successfully built 1b1278345097
Successfully tagged swift-nio-ssl:18.04-5.0
tomerd@tombp swift-nio-ssl (master) $ docker run -it -v `pwd`:/code -w /code swift-nio-ssl:18.04-5.0 bash -l
root@c60fded33432:/code# swift --version
Swift version 5.0-dev (LLVM 9a8bf9ce12, Clang eba26b8d1c, Swift b74d54a27c)
Target: x86_64-unknown-linux-gnu
root@c60fded33432:/code# head Package.swift 
// swift-tools-version:5.0
//===----------------------------------------------------------------------===//
//
// This source file is part of the SwiftNIO open source project
//
// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
// Licensed under Apache License v2.0
//
// See LICENSE.txt for license information
// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
root@c60fded33432:/code# swift build 
/code: error: package at '/code' is using Swift tools version 4.2.0 which is no longer supported; use 5.0.0 or newer instead
root@c60fded33432:/code#
```

**Comment by tomerd** [comment]
@swift-nio-bot test this please

**Comment by Lukasa** [comment]
Fixed by #66.

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Comment by weissi** [comment]
@swift-nio-bot test this please

File: Sources/NIOOpenSSL/ByteBufferBIO.swift [hunk 1]
```diff
@@ -313,7 +313,7 @@ final class ByteBufferBIO {
             assert(!self.mustClearOutboundBuffer)
         }
 
-        let writtenBytes = self.outboundBuffer.write(bytes: buffer)
+        let writtenBytes = self.outboundBuffer.writeBytes(buffer)
```
Comment (weissi) [inline Sources/NIOOpenSSL/ByteBufferBIO.swift]: @tanner0101 this should `import _NIO1APIShims` so it doesn't get broken all the time

Comment (tanner0101) [reply Sources/NIOOpenSSL/ByteBufferBIO.swift]: done: https://github.com/apple/swift-nio-ssl/pull/60/commits/559bda3d084ac60d606e7478393a4f1fa2acb8ed

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Comment by weissi** [comment]
that has been done as part of #59

**Pull request closed without merging**